gradle-plugin/src/main/java/com/google/cloud/tools/dependencies/gradle/LinkageCheckTask.java

```diff
@@ -191,59 +197,162 @@ private String dependencyPathsOfProblematicJars(
-            (ResolvedDependencyResult) dependencyResult;
-        ResolvedComponentResult child = resolvedDependencyResult.getSelected();
+    @Override
+    public boolean equals(Object other) {
+      if (this == other) {
+        return true;
+      }
+      if (other == null || getClass() != other.getClass()) {
+        return false;
+      }
+      ResolvedComponentResultNode that = (ResolvedComponentResultNode) other;
+      return Objects.equals(componentResult, that.componentResult)
+          && Objects.equals(parent, that.parent);
+    }
 
-        if (stack.contains(child)) {
-          // Circular dependency check
-          if (checkedCircularDependency.add(child)) {
+    @Override
+    public int hashCode() {
+      return Objects.hash(componentResult, parent);
+    }
+  }
+
+  /**
+   * Returns mapping from Maven coordinates to their dependency paths appearing in the dependency
+   * graph
+   *
+   * @param componentResult The root project in the dependency graph
+   * @param targetCoordinatesSet The Maven coordinates to check their dependency paths
+   */
+  private ListMultimap<String, String> groupCoordinatesToDependencyPaths(
+      ResolvedComponentResult componentResult, Set<String> targetCoordinatesSet) {
+
+    ListMultimap<String, String> coordinatesToDependencyPaths =
+        MultimapBuilder.hashKeys().arrayListValues().build();
+    // No need to print the same circular dependency multiple times
+    Set<ResolvedComponentResult> checkedCircularDependency = new HashSet<>();
+
+    for (String targetCoordinates : targetCoordinatesSet) {
+      // Queue of dependnecy nodes. Each node knows its parent.
+      ArrayDeque<ResolvedComponentResultNode> queue = new ArrayDeque<>();
+      ResolvedComponentResultNode firstItem =
+          new ResolvedComponentResultNode(componentResult, null);
+      queue.add(firstItem);
+
+      // Mapping to omit duplicate dependency paths in the output. This is a mapping from Maven
+      // coordinates to dependency nodes that has the dependency of the coordinate in their direct
+      // or transitive dependencies.
+      Set<ResolvedComponentResult> nodesDependOnTarget = new HashSet<>();
+
+      while (!queue.isEmpty()) {
+        ResolvedComponentResultNode node = queue.poll();
+        ResolvedComponentResult item = node.componentResult;
+
+        ModuleVersionIdentifier identifier = item.getModuleVersion();
+        String coordinates =
+            String.format(
+                "%s:%s:%s", identifier.getGroup(), identifier.getName(), identifier.getVersion());
+        if (targetCoordinates.equals(coordinates)) {
+          String dependencyPath = node.pathFromRoot();
+          coordinatesToDependencyPaths.put(targetCoordinates, dependencyPath);
+
+          for (ResolvedComponentResultNode iter = node.parent; iter != null; iter = iter.parent) {
+            nodesDependOnTarget.add(iter.componentResult);
+          }
+        }
+
+        if (nodesDependOnTarget.contains(item)) {
```

Comment:
Should this `if` block go above the ` if (targetCoordinates.equals(coordinates)) {`? What if a transitive dependency on `targetCoordinates` is found before the direct dependency? Or is it a "this should never happen in a gradle resolution" type of situation?

Reply:
If `nodesDependOnTarget.contains(item)` is true, then targetCoordinates (the coordinates of `item`) is not equal to coordinates.

> What if a transitive dependency on targetCoordinates is found before the direct dependency?

Would you explain an example?

Reply:
I was thinking it's possible to have a duplicate if both  ` if (targetCoordinates.equals(coordinates)) ` from ~10 lines above and this current if statement are both true.

Reply:
The 2 if-statements don't become true at the same time. I just updated this to the below to deliver a more clear meaning.

```
        if (targetCoordinates.equals(coordinates)) {
          ...
        } else if (nodesDependOnTarget.contains(item)) {
          ...
        } else {
          queue.addAll(getDependencies(node));
        }
```